### ISSUE TITLE
File upload not working in production

### ISSUE THREAD

**Issue by lsliwaradioluz** [strapi/strapi#5259]
**Describe the bug**
In my Strapi app, I am hosting my files with Cloudinary. I upload them via my frontend, using nuxt-apollo queries. While testing on localhost, the file upload works fine. When I test in in production, I get an error related to CORS.

Important! The bug appeared after moving from beta 18.5 to beta 18.7. Before it was non-existent. 

All other mutations and queries work fine in both development and production. 

Important! The bug appeared after moving from beta 18.5 to beta 18.7. Before it was non-existent. 

**Expected behavior**
Get 200 response after attempting to upload the file with nuxt-apollo

**Screenshots**

![image](https://user-images.githubusercontent.com/39009379/74735351-cc7fa480-5250-11ea-91f0-353de297ef11.png)

**Code snippets**

This is my config/environments/production/security.json file:

`{
  "csp": {
    "enabled": true,
    "policy": ["block-all-mixed-content"]
  },
  "p3p": {
    "enabled": true,
    "value": ""
  },
  "hsts": {
    "enabled": true,
    "maxAge": 31536000,
    "includeSubDomains": true
  },
  "xframe": {
    "enabled": true,
    "value": "SAMEORIGIN"
  },
  "xss": {
    "enabled": true,
    "mode": "block"
  },
  "cors": {
    "enabled": true, 
    "origin": ["https://www.piti.live","https://powerful-taiga-81942.herokuapp.com"]
  },
  "ip": {
    "enabled": false,
    "whiteList": [],
    "blackList": []
  }
}
`

**System**
- Strapi version: beta 18.7
- Database: mongodb
- Operating system: Windows 10

**Additional context**
This might be related to #2558 and #1919 since users were getting same errors there. However - if I understood correctly - they were getting this errors on localhost as well, whereas mine only happen in production.

**Comment by lauriejim** [comment]
Hello! Can you try with a fresh new app, just to know if it's a migration issue or related to Strapi itself?

**Comment by lsliwaradioluz** [comment]
Ok @lauriejim , I'll give it a go and let you know

**Comment by lsliwaradioluz** [comment]
@lauriejim I did what you'd asked for. I created a brand new Strapi backend (beta 18.7) and connected it to Nuxt.js front which is using nuxt-apollo. Both are deployed to Heroku. 

The exact same thing happens: on localhost file upload and all mutations work fine, whereas in production attempting to upload a file throws out the same CORS error. What shall I do next?

**Comment by lauriejim** [comment]
OKay, in your production server you start you app in `production` environment.
CORS are activated in this env.
See `./config/environments/production/security.json`
And it is no in development.
Here is the docs to configure your cors settings - https://strapi.io/documentation/3.0.0-beta.x/concepts/configurations.html#security

**Comment by jeremy-scroggs** [comment]
I'm experiencing the same issue on fresh install of: 
strapi 3.0.0-beta.18.8
sqlite3 latest
node 12.0.0

I have a form that takes in a bunch of data and some media. Two of the fields are media. One is a singular media object and the other is a collection. On local, all upload calls work as expected. On production, the singular media object upload works as expected, while the media collection is the upload request that fails. 

In the code below, you can see each of the calls which are almost identical:

**Single File Upload**
```
function SubmitPrimary(entryId: string){
    let formData = new FormData();
    const request = new XMLHttpRequest();

    formData.append("ref", "entry");
    formData.append("refId", entryId);
    formData.append("field", "primary");
    formData.append("files", primaryInputFile ? primaryInputFile : '');
    
    request.open('POST', process.env.REACT_APP_PROD_BASE_URI + '/upload');
    request.send(formData);
  }
```
**Multiple File Upload**
```
  function SubmitSupplementary(entryId: string){
    if(supplementaryFiles){
      let formData = new FormData();
      const request = new XMLHttpRequest();

      formData.append("ref", "entry");
      formData.append("refId", entryId);
      formData.append("field", "supplementary");
      supplementaryFiles.forEach((s) => {
        if(s.file)
          formData.append('files', s.file, s.file.name);
      })
      
      request.open('POST', process.env.REACT_APP_PROD_BASE_URI + '/upload');
      request.send(formData);
    }
  }
```
Additionally, I have checked that my app is started in production environment and that the cors origin is update to "*". 

Let me know if there is any other information that might help pin this down.

**Comment by noqcks** [comment]
I am also unable to use file upload in production.

Getting the error
```
Access to XMLHttpRequest at 'http://<server-ip>:1338/upload' from origin 'http://<server-ip>:3000' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.
```

In strapi logs I see
```
[2020-04-28T14:22:57.346Z] debug OPTIONS /upload (0 ms) 204
  AssertionError [ERR_ASSERTION]: Cannot wrap non-Error object
      at Function.boomify (/home/ec2-user/data/noqcks/EQX/backend/node_modules/boom/lib/index.js:107:14)
      at formatBoomPayload (/home/ec2-user/data/noqcks/EQX/backend/node_modules/strapi/lib/middlewares/boom/index.js:47:22)
      at /home/ec2-user/data/noqcks/EQX/backend/node_modules/strapi/lib/middlewares/boom/index.js:84:36
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (internal/process/task_queues.js:97:5)
```

My frontend code looks like this
```
    const formData = new FormData();
    formData.append("files", e.target.files[0]);
    axios
      .post(apiUrl + '/upload', formData, {
        headers: {
          'Authorization': 'Bearer ' + Cookies.get("jwt"),
          'Content-Type': 'multipart/form-data'
        }
      })
```

And my security settings look like this 
```
  "cors": {
    "enabled": true,
    "origin": "*"
  },
```

Any ideas @lauriejim?

**Comment by noqcks** [comment]
> Update: I've also tried uploading using fetch() instead of apollo mutation and it works! So I guess it is not a problem with Strapi?


@lsliwaradioluz can you post how you are making the fetch() request? Thanks

**Comment by lucasgehin** [comment]
Hello,
Any update on this ?
I'm facing exactly the same problem, I have a CORS issue with upload but every other request works perfectly
Thanks

**Comment by lauriejim** [comment]
Hello @lucasgehin can you please create a new issue - it will be easier to answer.

This issue is now too old and I have no update.

Make sure you well configured your CORS settings - https://strapi.io/documentation/3.0.0-beta.x/concepts/configurations.html#security
